--- Sendmail.pm.orig	2010-01-07 17:28:36.000000000 +0700
+++ Sendmail.pm	2010-01-07 17:36:06.000000000 +0700
@@ -9,7 +9,7 @@
 
 =cut
 
-$VERSION = '0.79';
+$VERSION = '0.791'; # patched to allow setting envelope sender
 
 # *************** Configuration you may want to change *******************
 # You probably want to set your SMTP server here (unless you specify it in
@@ -22,6 +22,8 @@
 
     'from'    => '', # default sender e-mail, used when no From header in mail
 
+    'sender'  => '', # default envelope sender address
+
     'mime'    => 1, # use MIME encoding by default
 
     'retries' => 1, # number of retries on smtp connect failure
@@ -122,7 +124,7 @@
 
     my (%mail, $k,
         $smtp, $server, $port, $connected, $localhost,
-        $fromaddr, $recip, @recipients, $to, $header,
+        $sender, $fromaddr, $recip, @recipients, $to, $header,
        );
 
     # -------- a few internal subs ----------
@@ -213,12 +215,14 @@
     # Extract 'From:' e-mail address to use as envelope sender
 
     $fromaddr = $mail{'Sender'} || $mail{'From'} || $mailcfg{'from'};
-    delete $mail{'Sender'};
     unless ($fromaddr =~ /$address_rx/) {
         return fail("Bad or missing From address: \'$fromaddr\'");
     }
     $fromaddr = $1;
 
+    $sender = $mail{'Env_Sender'} || $mail{'Sender'} || $mailcfg{'sender'} || $fromaddr;
+    delete $mail{'Sender'};
+
     # add Date header if needed
     $mail{Date} ||= time_to_date() ;
     $log .= "Date: $mail{Date}\n";
@@ -332,7 +336,7 @@
         || return fail("send HELO error");
     socket_read()
         || return fail("HELO error ($_)");
-    socket_write("MAIL FROM: <$fromaddr>$CRLF")
+    socket_write("MAIL FROM: <$sender>$CRLF")
         || return fail("send MAIL FROM: error");
     socket_read()
         || return fail("MAIL FROM: error ($_)");
